File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Hard/mergeOverlappingSubintervals.ts
function mergeOverlappingSubintervals(arr: number[][]) {
	const ans: number[][] = [];
	arr.sort((a, b) => a[0] - b[0]);
	for (let i = 0; i < arr.length; i++) {
		if (ans.length === 0 || ans[ans.length - 1][1] < arr[i][0]) {
			ans.push(arr[i]);
		} else {
			ans[ans.length - 1][1] = Math.max(ans[ans.length - 1][1], arr[i][1]);
		}
	}
	console.log(ans);
}

mergeOverlappingSubintervals([[1, 3], [2, 6], [8, 9], [9, 11], [8, 10], [2, 4], [15, 18], [16, 17]]);

File: StriverA2Z/Java/Arrays/Hard/MergeOverlappingSubintervals.java
import java.util.*;

public class MergeOverlappingSubintervals {
    public static void main(String[] args) {
        int[][] arr = {{1, 3}, {2, 6}, {8, 9}, {9, 11}, {8, 10}, {2, 4}, {15, 18}, {16, 17}};
        mergeOverlappingSubintervals(arr);
    }

    static void mergeOverlappingSubintervals(int[][] arr) {
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(arr, new Comparator<int[]>() {
            @Override
            public int compare(int[] row1, int[] row2) {
                // Compare based on the first element of each row
                return Integer.compare(row1[0], row2[0]);
            }
        });
        for (int i = 0; i < arr.length; i++) {
            if (ans.isEmpty() || ans.get(ans.size() - 1).get(1) < arr[i][0]) {
                List<Integer> temp = new ArrayList<>();
                temp.add(arr[i][0]);
                temp.add(arr[i][1]);
                ans.add(temp);
            } else {
                ans.get(ans.size() - 1).set(1, Math.max(ans.get(ans.size() - 1).get(1), arr[i][1]));
            }
        }

        for (List<Integer> i : ans) {
            System.out.println(i);
        }
    }
}
